executor/write.go

```diff
@@ -45,6 +46,76 @@ var (
 	_ exec.Executor = &LoadDataExec{}
 )
 
+// exchangePartitionCheckRow is only used for ExchangePartition durating write only state.
+// It check if rowData inserted or updated violate partition definition or check constraints.
+func exchangePartitionCheckRow(sctx sessionctx.Context, row []types.Datum, t table.Table) error {
+	tbl := t.Meta()
+	if tbl.ExchangePartitionInfo.CurrentIsPartitionTable == false {
+		is := sctx.GetDomainInfoSchema().(infoschema.InfoSchema)
+		pt, tableFound := is.TableByID(tbl.ExchangePartitionInfo.ExchangePartitionTableID)
+		if !tableFound {
+			return errors.Errorf("exchange partition process table by id failed")
+		}
+		p, ok := pt.(table.PartitionedTable)
+		if !ok {
+			return errors.Errorf("exchange partition process assert table partition failed")
+		}
+		err := p.CheckForExchangePartition(
+			sctx,
+			pt.Meta().Partition,
+			row,
+			tbl.ExchangePartitionInfo.ExchangePartitionPartitionID,
+		)
+		if err != nil {
+			return err
+		}
+		if variable.EnableCheckConstraint.Load() {
+			cc, ok := pt.(table.CheckConstraintTable)
+			if !ok {
+				return errors.Errorf("exchange partition process assert check constraint failed")
+			}
+			err := cc.CheckRowConstraint(sctx, row)
+			if err != nil {
+				// TODO: make error include ExchangePartition info.
+				return err
+			}
+		}
+	} else {
+		if variable.EnableCheckConstraint.Load() {
+			p, ok := t.(table.PartitionedTable)
+			if !ok {
+				return errors.Errorf("exchange partition process assert table partition failed")
+			}
+			physicalTable, err := p.GetPartitionByRow(sctx, row)
```

Comment:
If this was called from `AddRecord()`/`UpdateRecord()` instead of the `fillRow()`/`updateRecord()`, then the partition pruning would already been done. Could that be a follow-up PR?

Reply:
I'm not very sure about what you said. Could you explain more about it?

Reply:
Hi and sorry for the delay. Here is the stack from this line:
![Screenshot 2023-09-08 at 16 45 16](https://github.com/pingcap/tidb/assets/5520054/d65203ad-3ed2-40b8-b80b-1cd2d68dc4cd)
And later in AddRecord locatePartition is called again:
![Screenshot 2023-09-08 at 16 45 37](https://github.com/pingcap/tidb/assets/5520054/ddb72dc0-4ce5-4653-886b-592e4a869962)
Which then results in the constraint check and insert (similar with update/deletes) will be applied to the physical partition here:
![Screenshot 2023-09-08 at 16 49 44](https://github.com/pingcap/tidb/assets/5520054/fda1d225-0f71-4d59-96e4-c64b2d9ff021)

So I'm just curious if we should change the place in the code where these checks for Exchange Partition is executed, to somewhere where we already have the located partition?
But we can fix that in a followup PR, in that case.

Reply:
Good advice, you really showed strong experience in this. I'm agree with you, and fix it in this pr together.